lib/semantics/assignment.cc

```diff
@@ -26,210 +26,231 @@
+  }
+  void Check(const Expr<SomeType> &);
+  void Check(const NullPointer &) {}  // P => NULL() without MOLD=; always OK
+
+  template<typename T> void Check(const FunctionRef<T> &f) {
+    const Symbol *ultimate{nullptr};
+    std::string funcName;
+    if (const auto *symbol{f.proc().GetSymbol()}) {
+      funcName = symbol->name().ToString();
+      ultimate = &symbol->GetUltimate();
+    } else if (const auto *intrinsic{f.proc().GetSpecificIntrinsic()}) {
+      funcName = intrinsic->name;
+    }
+    if (auto proc{
+            characteristics::Procedure::Characterize(f.proc(), intrinsics_)}) {
+      std::optional<parser::MessageFixedText> error;
+      if (const auto &funcResult{proc->functionResult}) {  // C1025
+        const auto *frProc{funcResult->IsProcedurePointer()};
+        if (procedure_.has_value()) {
+          // Shouldn't be here in this function unless lhs
+          // is an object pointer.
+          error =
+              "Procedure %s is associated with the result of a reference to function '%s' that does not return a procedure pointer"_err_en_US;
+        } else if (frProc != nullptr) {
+          error =
+              "Object %s is associated with the result of a reference to function '%s' that is a procedure pointer"_err_en_US;
+        } else if (!funcResult->attrs.test(
+                       characteristics::FunctionResult::Attr::Pointer)) {
+          error =
+              "%s is associated with the result of a reference to function '%s' that is a not a pointer"_err_en_US;
+        } else if (isContiguous_ &&
+            !funcResult->attrs.test(
+                characteristics::FunctionResult::Attr::Contiguous)) {
+          error =
+              "CONTIGUOUS %s is associated with the result of reference to function '%s' that is not contiguous"_err_en_US;
+        } else if (type_) {
+          const auto *frTypeAndShape{funcResult->GetTypeAndShape()};
+          CHECK(frTypeAndShape != nullptr);
+          if (!type_->IsCompatibleWith(messages_, *frTypeAndShape)) {
+            error =
+                "%s is associated with the result of a reference to function '%s' whose pointer result has an incompatible type or shape"_err_en_US;
+          }
         }
+      } else {
+        error =
+            "%s is associated with the non-existent result of reference to procedure"_err_en_US;
       }
-    } else {
-      error = "Pointer was assigned the non-existent "
-              "result of reference to procedure"_err_en_US;
-    }
-    if (error.has_value()) {
-      if (auto *msg{messages.Say(*error, lhs.name(), funcName)}) {
-        msg->Attach(lhs.name(), "Declaration of pointer"_en_US);
-        if (ultimate != nullptr) {
-          msg->Attach(ultimate->name(), "Declaration of function"_en_US);
+      if (error.has_value()) {
+        if (auto *msg{Say(*error, description_, funcName)}) {
+          if (ultimate != nullptr) {
+            msg->Attach(ultimate->name(), "Declaration of function"_en_US);
```

Comment:
I don't think this will have a source location if `ultimate` came from a `.mod` file.

Reply:
I'll look into it.  If it's a problem here, it's a problem at other sites as well.

Reply:
`SayAlreadyDeclared` in `resolve-names.cc` has one solution to this problem: If the symbol is not use-associated it shows the declaration at the source location of the name. Otherwise it reports the module it was use-associated from.